### PR TITLE
WHO 6 cards

### PR DESCRIPTION
**Pull request by Simisays**

### CODE REVIEW

File: forge-gui/res/cardsfolder/upcoming/tardis.txt [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+T:Mode$ Attacks | ValidCard$ Card.Self | IsPresent$ Time-Lord.YouCtrl | TriggerZones$ Battlefield | Execute$ TrigEffect | TriggerDescription$ Whenever CARDNAME attacks, if you control a Time Lord, the next spell you cast this turn has cascade and you may planeswalk.
+SVar:TrigEffect:DB$ Effect | StaticAbilities$ GrantCascade | Triggers$ ExileEffect | SubAbility$ DBPlaneswalk
+SVar:DBPlaneswalk:DB$ Planeswalk | Optional$ True
+SVar:GrantCascade:Mode$ Continuous | EffectZone$ Command | Affected$ Card.YouCtrl | AffectedZone$ Stack | AddKeyword$ Cascade | Description$ The next spell you cast this turn has cascade and you may planeswalk.
+SVar:ExileEffect:Mode$ SpellCast | EffectZone$ Command | ValidCard$ Card.YouCtrl | Execute$ RemoveEffect | Static$ True
+SVar:RemoveEffect:DB$ ChangeZone | Origin$ Command | Destination$ Exile | Defined$ Self
+K:Crew:2
+DeckNeeds:Type$Time Lord
+SVar:HasAttackEffect:TRUE
+Oracle:Flying\nWhenever TARDIS attacks, if you control a Time Lord, the next spell you cast this turn has cascade and you may planeswalk.\nCrew 2 (Tap any number of creatures you control w
```
Comment (tool4ever) [inline forge-gui/res/cardsfolder/upcoming/tardis.txt:15]: incomplete

File: forge-gui/res/cardsfolder/upcoming/the_doctors_tomb.txt [hunk 2]
```diff
@@ -0,0 +1,10 @@
+Name:The Doctor's Tomb
+ManaCost:no cost
+Types:Plane Trenzalore
+R:Event$ Moved | ActiveZones$ Command | Destination$ Graveyard | ValidCard$ Card.Creature | ReplaceWith$ Exile | Description$ If a creature would die, instead exile it and that creature's controller loses 2 life.
+SVar:Exile:DB$ ChangeZone | Hidden$ True | Origin$ All | Destination$ Exile | Defined$ ReplacedCard | SubAbility$ DBLoseLife
+SVar:DBLoseLife:DB$ LoseLife | Defined$ CardController | LifeAmount$ 2
+T:Mode$ ChaosEnsues | TriggerZones$ Command | Execute$ TrigRedistribute | TriggerDescription$ Whenever chaos ensues, redistribute any number of players' life totals. (Each of those players gets one life total back.)
+SVar:TrigRedistribute:DB$ SetLife | PlayerChoices$ Player | ChoiceAmount$ Any | ChoicePrompt$ Choose any number of players | Redistribute$ True
+SVar:ETBPrevent:Event$ Moved | ActiveZones$ Battlefield | Origin$ Exile | Destination$ Battlefield | ValidLKI$ Card | Prevent$ True | Layer$ CantHappen | Description$ cards can't enter the battlefield from exile this turn.
```
Comment (tool4ever) [inline forge-gui/res/cardsfolder/upcoming/the_doctors_tomb.txt:9]: ?

File: forge-game/src/main/java/forge/game/ability/effects/PlaneswalkEffect.java [hunk 3]
```diff
@@ -24,9 +25,16 @@ public void resolve(SpellAbility sa) {
         }
         if (sa.hasParam("Defined")) {
             CardCollectionView destinations = AbilityUtils.getDefinedCards(sa.getHostCard(), sa.getParam("Defined"), sa);
-            sa.getActivatingPlayer().planeswalkTo(sa, destinations);
+            sa.getActivatingPlayer().planeswalkTo(sa, destinations);  
+        }
+    	{
+        for (Player p : game.getPlayers())
```
[the remark this answers is not in the document]
Comment (tool4ever) [reply forge-game/src/main/java/forge/game/ability/effects/PlaneswalkEffect.java]: imo the logic still looks weird, you could just have L32-34 before the Defined and avoid further changes

File: forge-gui/res/cardsfolder/upcoming/exterminate.txt [hunk 4]
```diff
@@ -0,0 +1,8 @@
+Name:Exterminate!
+ManaCost:2 B
+Types:Sorcery
+K:Replicate:tapXType<1/Dalek>
+A:SP$ Destroy | Cost$ 1 B B | ValidTgts$ Creature | SubAbility$ DBDrain | SpellDescription$ Destroy target creature. Its controller loses 3 life.
```
Comment (Northmoc) [inline forge-gui/res/cardsfolder/upcoming/exterminate.txt:5]: Cost in this line doesn't match line 2

File: forge-gui/res/languages/de-DE.properties [hunk 5]
```diff
@@ -3297,3 +3297,4 @@ lblWurm=Wurm
 lblYeti=Yeti
 lblZombie=Zombie
 lblZubera=Zubera
+lblWouldYouLikeToPlaneswalk=Would you like to Planeswalk ?
```
Comment (Northmoc) [inline forge-gui/res/languages/de-DE.properties:3300]: There should be an area for Effect prompts... the Effects are organized alphabetically. You could look for PlayEffect and put the stuff for Planeswalk effect just above

File: forge-game/src/main/java/forge/game/ability/effects/PlaneswalkEffect.java [hunk 6]
```diff
@@ -24,9 +25,16 @@ public void resolve(SpellAbility sa) {
         }
         if (sa.hasParam("Defined")) {
             CardCollectionView destinations = AbilityUtils.getDefinedCards(sa.getHostCard(), sa.getParam("Defined"), sa);
-            sa.getActivatingPlayer().planeswalkTo(sa, destinations);
+            sa.getActivatingPlayer().planeswalkTo(sa, destinations);  
+        }
+    	{
+         for ( Player p : getTargetPlayers(sa))
```
Comment (Northmoc) [inline forge-game/src/main/java/forge/game/ability/effects/PlaneswalkEffect.java:31]: `{` should be after for statement?

doesn't this crash your forge?

File: forge-gui/res/cardsfolder/upcoming/tardis.txt [hunk 7]
```diff
@@ -0,0 +1,15 @@
+Name:TARDIS
+ManaCost:2
+Types:Artifact Vehicle
+PT:2/4
+K:Flying
+T:Mode$ Attacks | ValidCard$ Card.Self | IsPresent$ Card.Time-Lord+YouCtrl | TriggerZones$ Battlefield | Execute$ TrigEffect | TriggerDescription$ Whenever CARDNAME attacks, if you control a Time Lord, the next spell you cast this turn has cascade and you may planeswalk.
+SVar:TrigEffect:DB$ Effect | StaticAbilities$ GrantCascade | Triggers$ ExileEffect | SubAbility$ DBPlaneswalk
+SVar:DBPlaneswalk:DB$ Planeswalk | Optional$ True
+SVar:GrantCascade:Mode$ Continuous | EffectZone$ Command | Affected$ Card.YouCtrl | AffectedZone$ Stack | AddKeyword$ Cascade | Description$ The next spell you cast this turn has cascade and you may planeswalk.
```
Comment (Northmoc) [inline forge-gui/res/cardsfolder/upcoming/tardis.txt:9]: I think `Description$ The next spell you cast this turn has cascade.` Since you'll have dealt with the planeswalk already by the time a human player could inspect the effect

File: forge-gui/res/lists/TypeLists.txt [hunk 8]
```diff
@@ -273,6 +274,7 @@ Thalakos:Thalakoses
 Thopter:Thopters
 Thrull:Thrulls
 Tiefling:Tieflings
+Time-Lord:Time Lords
```
Comment (Northmoc) [inline forge-gui/res/lists/TypeLists.txt:277]: Is this dealt with elsewhere in the code to get it to display right?

File: forge-gui/res/languages/es-ES.properties [hunk 9]
```diff
@@ -3300,3 +3300,4 @@ lblWurm=Wurm
 lblYeti=Yeti
 lblZombie=Zombi
 lblZubera=Zubera
+lblWouldYouLikeToPlaneswalk=Would you like to Planeswalk ?
```
Comment (Northmoc) [inline forge-gui/res/languages/es-ES.properties:3303]: I'd do `Would you like to planeswalk?` (no cap or extra space)

File: forge-gui/res/cardsfolder/upcoming/tardis_bay.txt [hunk 10]
```diff
@@ -0,0 +1,9 @@
+Name:TARDIS Bay
+ManaCost:no cost
+Types:Plane Gallifrey
+S:Mode$ Continuous | Affected$ Card.YouCtrl+cmcGE2 | EffectZone$ Command | Condition$ PlayerTurn | AffectedZone$ Stack | AddKeyword$ Cascade | CheckSVar$ Y | SVarCompare$ EQ0 | Description$ The first spell you cast during each of your turns with mana value 2 or greater has cascade. (When you cast that spell, exile cards from the top of your library until you exile a nonland card that costs less. You may cast it without paying its mana cost. Put the exiled cards on the bottom in a random order.)
+SVar:Y:Count$ThisTurnCast_Card.YouCtrl+cmcGE2
+T:Mode$ ChaosEnsues | TriggerZones$ Command | Execute$ TrigSteal | TriggerDescription$ When chaos ensues, gain control of target artifact. Then planeswalk.
+SVar:TrigSteal:DB$ GainControl | TgtPrompt$ Select target Artifact | ValidTgts$ Artifact | SubAbility$ DBPlaneswalk
```
Comment (Northmoc) [inline forge-gui/res/cardsfolder/upcoming/tardis_bay.txt:7]: Probably just remove that TgtPrompt

File: forge-gui/res/languages/es-ES.properties (hunk 9 above)
Comment (Northmoc) [reply forge-gui/res/languages/es-ES.properties:3303]: Oops in English anyway... sometimes I get en-US and es-ES mixed. Other languages probably matter less  to fix in that way, as they're just placeholders.

File: forge-gui/res/lists/TypeLists.txt (hunk 8 above)
Comment (Simisays) [reply forge-gui/res/lists/TypeLists.txt:277]: Don't know where to do that. Only this way can I get the IsPresent of TARDIS to work.

File: forge-gui/res/languages/es-ES.properties (hunk 9 above)
Comment (Simisays) [reply forge-gui/res/languages/es-ES.properties:3303]: Yeah I also feel like leaving this as the last entry since the people who do localizations can better reorganise these entries

File: forge-gui/res/cardsfolder/upcoming/tardis_bay.txt [hunk 11]
```diff
@@ -0,0 +1,9 @@
+Name:TARDIS Bay
+ManaCost:no cost
+Types:Plane Gallifrey
+S:Mode$ Continuous | Affected$ Card.YouCtrl+cmcGE2 | EffectZone$ Command | Condition$ PlayerTurn | AffectedZone$ Stack | AddKeyword$ Cascade | CheckSVar$ Y | SVarCompare$ EQ0 | Description$ The first spell you cast during each of your turns with mana value 2 or greater has cascade. (When you cast that spell, exile cards from the top of your library until you exile a nonland card that costs less. You may cast it without paying its mana cost. Put the exiled cards on the bottom in a random order.)
+SVar:Y:Count$ThisTurnCast_Card.YouCtrl+cmcGE2
+T:Mode$ ChaosEnsues | TriggerZones$ Command | Execute$ TrigSteal | TriggerDescription$ When chaos ensues, gain control of target artifact.
```
Comment (tool4ever) [inline forge-gui/res/cardsfolder/upcoming/tardis_bay.txt:6]: incomplete now

Comment (Northmoc) [reply forge-gui/res/cardsfolder/upcoming/tardis_bay.txt:6]: Yeah, I just meant to change the Desc on the Effect line!

File: forge-game/src/main/java/forge/game/ability/effects/PlaneswalkEffect.java [hunk 12]
```diff
@@ -18,13 +19,20 @@ public void resolve(SpellAbility sa) {
...
-                p.leaveCurrentPlane();
+            for (Player p : game.getPlayers()) {                
+            if (sa.hasParam("Optional") && !p.getController().confirmAction(sa, null,
+                Localizer.getInstance().getMessage("lblWouldYouLikeToPlaneswalk"), null)) {
+                    return;
+            }
+            else {              
+            	p.leaveCurrentPlane();
+            }
             }
```
Comment (Northmoc) [inline forge-game/src/main/java/forge/game/ability/effects/PlaneswalkEffect.java:30]: This formatting looks wonky

Line 23 should happen before we run the player loop... instead of using `p.getController()` it should probably be `sa.getActivatingPlayer().getController()`

File: forge-gui/res/cardsfolder/upcoming/exterminate.txt [hunk 13]
```diff
@@ -0,0 +1,8 @@
+Name:Exterminate!
+ManaCost:2 B
+Types:Sorcery
+K:Replicate:tapXType<1/Dalek>
+A:SP$ Destroy | ValidTgts$ Creature | SubAbility$ DBDrain | SpellDescription$ Destroy target creature. Its controller loses 3 life.
+SVar:DBDrain:DB$ LoseLife | Defined$ TargetedController | LifeAmount$ 3
```
Comment (Northmoc) [inline forge-gui/res/cardsfolder/upcoming/exterminate.txt:6]: OK, I know this is pedantic, but.

Say someone targets a creature with this. The desc goes to the stack. But then someone uses something at instant speed that results in the targeted creature changing controllers. The stack description for line 6 is now wrong.

So I think line 6 should be StackDescription$ SpellDescription and pull the second sentence from line 5 for the SpellDescription

Hopefully this makes sense 😅

File: forge-gui/res/cardsfolder/upcoming/tardis_bay.txt (hunk 11 above)
Comment (Northmoc) [reply forge-gui/res/cardsfolder/upcoming/tardis_bay.txt:6]: Just remember that Execute abilities and their subs don't append to TriggerDescription

File: forge-gui/res/languages/en-US.properties [hunk 14]
```diff
@@ -1870,6 +1870,7 @@ lblSearchPlayerZoneConfirm=Search {0} {1}?
 lblCardMatchSearchingTypeInAlternateZones=cards match your searching type in Alternate Zones.
 lblLookingCardIn=Looking at cards in
 lblDoYouWantPlayCard=Do you want to play {0}?
+lblWouldYouLikeToPlaneswalk=Would you like to planeswalk ?
```
Comment (Northmoc) [inline forge-gui/res/languages/en-US.properties:1873]: sorry, just `planeswalk?` not `planeswalk ?`

File: forge-gui/res/languages/es-ES.properties [hunk 15]
```diff
@@ -3299,4 +3299,4 @@ lblWraith=Fantasma
 lblWurm=Wurm
 lblYeti=Yeti
 lblZombie=Zombi
-lblZubera=Zubera
+lblZubera=Zubera
```
Comment (Northmoc) [inline forge-gui/res/languages/es-ES.properties:3302]: there should be a newline at the end, right?

and the new prompt should still be in the file for each language... I just meant to create a new spot for PlaneswalkEffect prompts

File: forge-game/src/main/java/forge/game/ability/effects/PlaneswalkEffect.java [hunk 16]
```diff
@@ -22,11 +23,14 @@ public void resolve(SpellAbility sa) {
...
+        }
         if (sa.hasParam("Defined")) {
             CardCollectionView destinations = AbilityUtils.getDefinedCards(sa.getHostCard(), sa.getParam("Defined"), sa);
-            sa.getActivatingPlayer().planeswalkTo(sa, destinations);
         } else {
             sa.getActivatingPlayer().planeswalk(sa);
         }
     }
-}
+}
```
Comment (Northmoc) [inline forge-game/src/main/java/forge/game/ability/effects/PlaneswalkEffect.java:36]: make sure we've got a blank line at the end

File: forge-game/src/main/java/forge/game/ability/effects/PlaneswalkEffect.java [hunk 17]
```diff
@@ -22,11 +23,14 @@ public void resolve(SpellAbility sa) {
                 p.leaveCurrentPlane();
             }
         }
+        if (sa.hasParam("Optional") && sa.getActivatingPlayer().getController().confirmAction(sa, null,
+                Localizer.getInstance().getMessage("lblWouldYouLikeToPlaneswalk"), null)) {
+                    return;
+        }
```
Comment (Northmoc) [inline forge-game/src/main/java/forge/game/ability/effects/PlaneswalkEffect.java:29]: We've already left because of the loop starting in line 22

I think you could just put this if statement right after the getActivePlanes if statement

File: forge-game/src/main/java/forge/game/ability/effects/PlaneswalkEffect.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (Northmoc) [reply forge-game/src/main/java/forge/game/ability/effects/PlaneswalkEffect.java]: This!

File: forge-gui/res/lists/TypeLists.txt (hunk 8 above)
Comment (Northmoc) [reply forge-gui/res/lists/TypeLists.txt:277]: I assume you tried changing both to "Time Lord" right?

What kind of error do you get?

File: forge-game/src/main/java/forge/game/ability/effects/PlaneswalkEffect.java [hunk 18]
```diff
@@ -22,11 +23,14 @@ public void resolve(SpellAbility sa) {
...
         if (sa.hasParam("Defined")) {
             CardCollectionView destinations = AbilityUtils.getDefinedCards(sa.getHostCard(), sa.getParam("Defined"), sa);
-            sa.getActivatingPlayer().planeswalkTo(sa, destinations);
```
Comment (Hanmac) [inline forge-game/src/main/java/forge/game/ability/effects/PlaneswalkEffect.java:27]: why is this line removed?

Comment (Simisays) [reply forge-game/src/main/java/forge/game/ability/effects/PlaneswalkEffect.java:27]: accident

File: forge-gui/res/lists/TypeLists.txt (hunk 8 above)
Comment (Simisays) [reply forge-gui/res/lists/TypeLists.txt]: nvm seems to work

File: forge-game/src/main/java/forge/game/ability/effects/PlaneswalkEffect.java [hunk 19]
```diff
@@ -16,17 +17,21 @@ public void resolve(SpellAbility sa) {
         if (game.getActivePlanes() == null) { // not a planechase game, nothing happens
             return;
         }
-
-        if (!sa.hasParam("DontPlaneswalkAway")) {
+        if (sa.hasParam("Optional") && sa.getActivatingPlayer().getController().confirmAction(sa, null,
+            Localizer.getInstance().getMessage("lblWouldYouLikeToPlaneswalk"), null)) {
+                return;
+        }
+        	if (!sa.hasParam("DontPlaneswalkAway")) {
```
Comment (Hanmac) [inline forge-game/src/main/java/forge/game/ability/effects/PlaneswalkEffect.java:24]: fix indent

File: forge-gui/res/cardsfolder/upcoming/time_lord_regeneration.txt [hunk 20]
```diff
@@ -0,0 +1,9 @@
...
+SVar:DiesTrig:Mode$ ChangesZone | Origin$ Battlefield | Destination$ Graveyard | ValidCard$ Card.Self | Execute$ TrigReveal | TriggerDescription$ When this creature dies, reveal cards from the top of your library until you reveal a Time Lord creature card. Put that card onto the battlefield and the rest on the bottom of your library in a random order.
+SVar:TrigReveal:DB$ DigUntil | Defined$ You | Valid$ Creature.Time Lord | FoundDestination$ Battlefield | RevealedDestination$ Library | RevealedLibraryPosition$ -1 | RevealRandomOrder$ True | StackDescription$ None
+DeckNeeds:Type$Time Lord|Changeling
+Oracle:Until end of turn, target Time Lord you control gains When this creature dies, reveal cards from the top of your library until you reveal a Time Lord creature card. Put that card onto the battlefield and the rest on the bottom of your library in a random order."
```
Comment (Northmoc) [inline forge-gui/res/cardsfolder/upcoming/time_lord_regeneration.txt:9]: Missing `"` also in lines 4 and 5

File: forge-game/src/main/java/forge/game/ability/effects/PlaneswalkEffect.java [hunk 21]
```diff
@@ -16,6 +17,10 @@ public void resolve(SpellAbility sa) {
         if (game.getActivePlanes() == null) { // not a planechase game, nothing happens
             return;
         }
+        if (sa.hasParam("Optional") && sa.getActivatingPlayer().getController().confirmAction(sa, null,
```
Comment (tool4ever) [inline forge-game/src/main/java/forge/game/ability/effects/PlaneswalkEffect.java:20]: for some reason you now inverted the logic

File: forge-gui/res/cardsfolder/upcoming/time_lord_regeneration.txt [hunk 22]
```diff
@@ -1,9 +1,9 @@
...
-SVar:Static:Mode$ Continuous | Affected$ Card.IsRemembered | EffectZone$ Command | AddTrigger$ DiesTrig | Description$ Until end of turn, target Time Lord you control gains When this creature dies, reveal cards from the top of your library until you reveal a Time Lord creature card. Put that card onto the battlefield and the rest on the bottom of your library in a random order."
+A:SP$ Animate | ValidTgts$ Card.Time Lord+YouCtrl | TgtPrompt$ Select target Time Lord you control | Triggers$ DiesTrig | SpellDescription$ Until end of turn, target Time Lord you control gains "When this creature dies, reveal cards from the top of your library until you reveal a Time Lord creature card. Put that card onto the battlefield and the rest on the bottom of your library in a random order."
```
Comment (Northmoc) [inline forge-gui/res/cardsfolder/upcoming/time_lord_regeneration.txt:4]: I think `ValidTgts$ Time Lord+YouCtrl` is a bit better

File: forge-gui/res/languages/zh-CN.properties [hunk 23]
```diff
@@ -3287,3 +3287,4 @@ lblWurm=亚龙
 lblYeti=雪怪
 lblZombie=灵俑
 lblZubera=无面鬼
+lblWouldYouLikeToPlaneswalk=Would you like to planeswalk?
```
Comment (Northmoc) [inline forge-gui/res/languages/zh-CN.properties:3290]: Something like this would be great for all languages
image

Comment (Northmoc) [reply forge-gui/res/languages/zh-CN.properties:3290]: You can see most effects are lumped together and alphabetical

File: forge-gui/res/languages/zh-CN.properties [hunk 24]
```diff
@@ -2042,7 +2042,9 @@ lblRevealCardToOtherPlayers=向其他玩家展示牌？
 lblChooseAnyNumberToPhase=选择任意要跃离的数量
 #PlayEffect.java
 lblChooseUpTo=最多选择
-lblSelectCardToPlay=选择要使用的牌
+lblSelectCardToPlay=选择要使用的�
```
Comment (tool4ever) [inline forge-gui/res/languages/zh-CN.properties:2045]: looks like this line shouldn't have been changed

File: forge-gui/res/cardsfolder/upcoming/tardis_bay.txt [hunk 25]
```diff
@@ -0,0 +1,9 @@
+Name:TARDIS Bay
+ManaCost:no cost
+Types:Plane Gallifrey
```
Comment (tool4ever) [inline forge-gui/res/cardsfolder/upcoming/tardis_bay.txt:3]: TypeLists not updated for planes?

**Review by tool4ever — APPROVED** [review]
